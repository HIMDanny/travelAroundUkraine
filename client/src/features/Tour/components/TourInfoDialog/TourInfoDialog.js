/* eslint-disable react/forbid-prop-types */

import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { useDispatch, useSelector } from 'react-redux';
import { styled, alpha } from '@mui/material/styles';
import Typography from '@mui/material/Typography';
import { Box as MuiBox, Button, IconButton, MenuItem as MuiMenuItem, Select as MuiSelect, Stack } from '@mui/material';
// Icons
import ExpandMoreIcon from '@mui/icons-material/ExpandMore';
import CalendarMonthIcon from '@mui/icons-material/CalendarMonth';
import AccessTimeIcon from '@mui/icons-material/AccessTime';
import FlightTakeoffIcon from '@mui/icons-material/FlightTakeoff';
import KeyboardReturnIcon from '@mui/icons-material/KeyboardReturn';
import CloseIcon from '@mui/icons-material/Close';

import formatTourDate from '../../utils/formatTourDate';
import currencyConverter from '../../utils/currencyConverter';
import convertData from '../../utils/convertData';

import { addProduct } from '../../../../store/slices/cartSlice/cartSlice';

const currencySymbol = {
  eur: '€',
  usd: '$',
  uah: '₴',
};

const BoxWrapper = styled(MuiBox)(({ theme }) => ({
  position: 'sticky',
  top: 20,
  padding: '20px 20px 30px',
  borderRadius: 5,

  '@media (min-width: 1051px)': {
    border: `1px solid ${theme.palette.divider}`,
  },
}));

const Section = styled(MuiBox)(({ theme }) => ({
  borderBottom: `1px solid ${theme.palette.divider}`,
  paddingBottom: 15,
  marginBottom: 20,
}));

const Title = styled((props) => <Typography variant="h3" color="text.primary" {...props} />)({
  textTransform: 'uppercase',
  fontWeight: 700,
  marginBottom: 20,
});

const IncludedContentWrapper = styled(MuiBox)({
  display: 'grid',
  gridTemplateColumns: 'repeat(auto-fit, minmax(160px, 1fr))',
  gap: '7px',
});

const Cost = styled(Typography)(({ theme }) => ({
  fontWeight: 700,
  marginBottom: 0,
  fontSize: 35,
  lineHeight: 1,
  color: theme.palette.primary.main,
}));

const Select = styled(MuiSelect)(({ theme }) => ({
  marginBottom: '10px',
  fontWeight: 700,
  '.MuiOutlinedInput-notchedOutline': {
    borderColor: `${theme.palette.divider}`,

    transition: theme.transitions.create('border-color'),
  },
  '&:hover .MuiOutlinedInput-notchedOutline': {
    borderColor: `${theme.palette.primary.main}`,
  },
  '.MuiSelect-icon': {
    transition: theme.transitions.create('transform'),
  },
}));

const MenuItem = styled(MuiMenuItem)(({ theme }) => ({
  transition: theme.transitions.create(['color', 'background-color']),

  '&:hover': {
    backgroundColor: alpha(theme.palette.primary.main, 0.15),
    color: theme.palette.primary.main,
  },

  '&.Mui-selected': {
    color: theme.palette.primary.main,
    fontWeight: 700,
  },
}));

const DetailsText = styled((props) => <Typography gutterBottom={false} {...props} />)({});

const TourInfoDialog = ({
  dates,
  professionalGuide,
  accommodation,
  meals,
  transferAlongTheRoute,
  travelInsurance,
  departs,
  duration,
  returns,
  currentPrice,
  id,
  itemNo,
  closeButton,
  handleClose,
}) => {
  const [currency, setCurrency] = useState('eur');
  const included = convertData(professionalGuide, accommodation, meals, transferAlongTheRoute, travelInsurance);
  const dispatch = useDispatch();
  const isLogin = useSelector((store) => store.userReducer.isLogin);

  return (
    <BoxWrapper>
      {closeButton && (
        <IconButton sx={{ position: 'absolute', right: '10px', top: '10px' }} onClick={handleClose}>
          <CloseIcon />
        </IconButton>
      )}
      <Section>
        <Title>Included</Title>

        <IncludedContentWrapper>
          {included.map(({ icon, service }) => (
            <Stack key={service} direction="row" gap="5px" alignItems="center">
              {icon}
              <Typography gutterBottom={false}>{service}</Typography>
            </Stack>
          ))}
        </IncludedContentWrapper>
      </Section>
      <Section>
        <Title>Dates</Title>
        <Stack direction="row" gap="5px" alignItems="center">
          <CalendarMonthIcon color="primary" />
          <Typography gutterBottom={false}>
            {formatTourDate(dates.beginDate)} - {formatTourDate(dates.endDate)}
          </Typography>
        </Stack>
      </Section>
      <Section>
        <Title>Details</Title>
        <Stack gap="5px">
          <Stack direction="row" alignItems="center" gap="5px">
            <AccessTimeIcon color="primary" />
            <DetailsText>Duration</DetailsText>
            <DetailsText marginLeft="auto">{duration}</DetailsText>
          </Stack>
          <Stack direction="row" alignItems="center" gap="5px">
            <FlightTakeoffIcon color="primary" />
            <DetailsText>Departs</DetailsText>
            <DetailsText marginLeft="auto">{departs}</DetailsText>
          </Stack>
          <Stack direction="row" alignItems="center" gap="5px">
            <KeyboardReturnIcon color="primary" />
            <DetailsText>Returns</DetailsText>
            <DetailsText marginLeft="auto">{returns}</DetailsText>
          </Stack>
        </Stack>
      </Section>
      <Section>
        <Title>Currency</Title>
        <Select fullWidth value={currency} onChange={(e) => setCurrency(e.target.value)} IconComponent={ExpandMoreIcon}>
          {Object.keys(currencySymbol).map((key) => (
            <MenuItem key={key} value={key}>
              {key.toUpperCase()}
            </MenuItem>
          ))}
        </Select>
      </Section>
      <MuiBox>
        <Title>Total</Title>
        <Stack direction="row" justifyContent="space-between" alignItems="center">
          <Cost>
            {currencySymbol[currency]}
            {currencyConverter(currentPrice, currency)}
          </Cost>
          <Button
            sx={{ paddingInline: '30px' }}
            disableElevation
            onClick={() => dispatch(addProduct(itemNo, id, isLogin))}
          >
            Add to Cart
          </Button>
        </Stack>
      </MuiBox>
    </BoxWrapper>
  );
};

TourInfoDialog.propTypes = {
  dates: PropTypes.objectOf(PropTypes.instanceOf(Date)).isRequired,
  professionalGuide: PropTypes.bool,
  accommodation: PropTypes.bool,
  meals: PropTypes.bool,
  transferAlongTheRoute: PropTypes.bool,
  travelInsurance: PropTypes.bool,
  duration: PropTypes.string,
  departs: PropTypes.string,
  returns: PropTypes.string,
  currentPrice: PropTypes.number,
  id: PropTypes.string,
  itemNo: PropTypes.string,
  closeButton: PropTypes.bool,
  handleClose: PropTypes.func,
};

TourInfoDialog.defaultProps = {
  closeButton: false,
  handleClose: () => {},
  professionalGuide: false,
  accommodation: false,
  meals: false,
  transferAlongTheRoute: false,
  travelInsurance: false,
  duration: 'Unknown',
  departs: 'Unknown',
  returns: 'Unknown',
  currentPrice: 0,
  id: '',
  itemNo: '',
};

export default TourInfoDialog;
