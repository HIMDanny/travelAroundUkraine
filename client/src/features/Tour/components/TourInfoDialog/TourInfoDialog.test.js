import { render } from '@testing-library/react';
import TourInfoDialog from './TourInfoDialog';

jest.mock('@mui/material/Box', () => ({ children }) => <div>{children}</div>);
jest.mock('@mui/material/Stack', () => ({ children }) => <div>{children}</div>);
jest.mock('@mui/material/Container', () => ({ children }) => <div>{children}</div>);
jest.mock('@mui/material/Typography', () => ({ children }) => <p>{children}</p>);
jest.mock('@mui/material/Button', () => ({ children }) => <button>{children}</button>);
jest.mock('@mui/material/IconButton', () => ({ children }) => <button>{children}</button>);
jest.mock('@mui/material/Select', () => ({ children }) => <select>{children}</select>);
jest.mock('@mui/material/MenuItem', () => ({ children }) => <option>{children}</option>);

const included = [
  { icon: '', service: 'Professional guide' },
  { icon: '', service: 'Accomodation' },
];

const cost = {
  eur: 70,
  usd: 70,
  uah: 2584,
};

const dates = { beginDate: new Date('2022-02-01'), endDate: new Date('2022-02-25') };

const details = {
  duration: '3 days',
  departs: 'First city',
  returns: 'Second city',
};

describe('TourInfoDialog snapshot testing', () => {
  test('should TourInfoDialog without close button match snapshot', () => {
    const { asFragment } = render(<TourInfoDialog included={included} cost={cost} dates={dates} details={details} />);

    expect(asFragment()).toMatchSnapshot();
  });

  test('should TourInfoDialog with close button match snapshot', () => {
    const { asFragment } = render(
      <TourInfoDialog included={included} cost={cost} dates={dates} details={details} closeButton />
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
