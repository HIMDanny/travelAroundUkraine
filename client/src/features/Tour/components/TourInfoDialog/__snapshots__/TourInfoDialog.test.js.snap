// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TourInfoDialog snapshot testing should TourInfoDialog with close button match snapshot 1`] = `
<DocumentFragment>
  <div>
    <div>
      <p>
        Included
      </p>
      <button>
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="CloseIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
          />
        </svg>
      </button>
      <div>
        <div>
          <p>
            Professional guide
          </p>
        </div>
        <div>
          <p>
            Accomodation
          </p>
        </div>
      </div>
    </div>
    <div>
      <p>
        Dates
      </p>
      <div>
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-colorPrimary MuiSvgIcon-fontSizeMedium css-ci5apu-MuiSvgIcon-root"
          data-testid="CalendarMonthIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 4h-1V2h-2v2H8V2H6v2H5c-1.11 0-1.99.9-1.99 2L3 20c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V6c0-1.1-.9-2-2-2zm0 16H5V10h14v10zM9 14H7v-2h2v2zm4 0h-2v-2h2v2zm4 0h-2v-2h2v2zm-8 4H7v-2h2v2zm4 0h-2v-2h2v2zm4 0h-2v-2h2v2z"
          />
        </svg>
        <p>
          2022.02.01 - 2022.02.25
        </p>
      </div>
    </div>
    <div>
      <p>
        Details
      </p>
      <div>
        <div>
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-colorPrimary MuiSvgIcon-fontSizeMedium css-ci5apu-MuiSvgIcon-root"
            data-testid="AccessTimeIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M11.99 2C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zM12 20c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
            />
            <path
              d="M12.5 7H11v6l5.25 3.15.75-1.23-4.5-2.67z"
            />
          </svg>
          <p>
            Duration
          </p>
          <p>
            3 days
          </p>
        </div>
        <div>
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-colorPrimary MuiSvgIcon-fontSizeMedium css-ci5apu-MuiSvgIcon-root"
            data-testid="FlightTakeoffIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M2.5 19h19v2h-19v-2zm19.57-9.36c-.21-.8-1.04-1.28-1.84-1.06L14.92 10l-6.9-6.43-1.93.51 4.14 7.17-4.97 1.33-1.97-1.54-1.45.39 2.59 4.49s7.12-1.9 16.57-4.43c.81-.23 1.28-1.05 1.07-1.85z"
            />
          </svg>
          <p>
            Departs
          </p>
          <p>
            First city
          </p>
        </div>
        <div>
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-colorPrimary MuiSvgIcon-fontSizeMedium css-ci5apu-MuiSvgIcon-root"
            data-testid="KeyboardReturnIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 7v4H5.83l3.58-3.59L8 6l-6 6 6 6 1.41-1.41L5.83 13H21V7z"
            />
          </svg>
          <p>
            Returns
          </p>
          <p>
            Second city
          </p>
        </div>
      </div>
    </div>
    <div>
      <p>
        Currency
      </p>
      <select>
        <option>
          EUR
        </option>
        <option>
          USD
        </option>
        <option>
          UAH
        </option>
      </select>
    </div>
    <div>
      <p>
        Total
      </p>
      <div>
        <p>
          €70
        </p>
        <button>
          Add to Cart
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`TourInfoDialog snapshot testing should TourInfoDialog without close button match snapshot 1`] = `
<DocumentFragment>
  <div>
    <div>
      <p>
        Included
      </p>
      <div>
        <div>
          <p>
            Professional guide
          </p>
        </div>
        <div>
          <p>
            Accomodation
          </p>
        </div>
      </div>
    </div>
    <div>
      <p>
        Dates
      </p>
      <div>
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-colorPrimary MuiSvgIcon-fontSizeMedium css-ci5apu-MuiSvgIcon-root"
          data-testid="CalendarMonthIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 4h-1V2h-2v2H8V2H6v2H5c-1.11 0-1.99.9-1.99 2L3 20c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V6c0-1.1-.9-2-2-2zm0 16H5V10h14v10zM9 14H7v-2h2v2zm4 0h-2v-2h2v2zm4 0h-2v-2h2v2zm-8 4H7v-2h2v2zm4 0h-2v-2h2v2zm4 0h-2v-2h2v2z"
          />
        </svg>
        <p>
          2022.02.01 - 2022.02.25
        </p>
      </div>
    </div>
    <div>
      <p>
        Details
      </p>
      <div>
        <div>
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-colorPrimary MuiSvgIcon-fontSizeMedium css-ci5apu-MuiSvgIcon-root"
            data-testid="AccessTimeIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M11.99 2C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zM12 20c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
            />
            <path
              d="M12.5 7H11v6l5.25 3.15.75-1.23-4.5-2.67z"
            />
          </svg>
          <p>
            Duration
          </p>
          <p>
            3 days
          </p>
        </div>
        <div>
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-colorPrimary MuiSvgIcon-fontSizeMedium css-ci5apu-MuiSvgIcon-root"
            data-testid="FlightTakeoffIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M2.5 19h19v2h-19v-2zm19.57-9.36c-.21-.8-1.04-1.28-1.84-1.06L14.92 10l-6.9-6.43-1.93.51 4.14 7.17-4.97 1.33-1.97-1.54-1.45.39 2.59 4.49s7.12-1.9 16.57-4.43c.81-.23 1.28-1.05 1.07-1.85z"
            />
          </svg>
          <p>
            Departs
          </p>
          <p>
            First city
          </p>
        </div>
        <div>
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-colorPrimary MuiSvgIcon-fontSizeMedium css-ci5apu-MuiSvgIcon-root"
            data-testid="KeyboardReturnIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 7v4H5.83l3.58-3.59L8 6l-6 6 6 6 1.41-1.41L5.83 13H21V7z"
            />
          </svg>
          <p>
            Returns
          </p>
          <p>
            Second city
          </p>
        </div>
      </div>
    </div>
    <div>
      <p>
        Currency
      </p>
      <select>
        <option>
          EUR
        </option>
        <option>
          USD
        </option>
        <option>
          UAH
        </option>
      </select>
    </div>
    <div>
      <p>
        Total
      </p>
      <div>
        <p>
          €70
        </p>
        <button>
          Add to Cart
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
